### DISCUSSION TITLE
Join errors in BoxableExpression

### DISCUSSION THREAD

**Discussion by SteelAlloy** [diesel-rs/diesel#3554] [Q&A] [answered]
Hi everyone,

I'm trying to construct a boolean expression in a function to filter values on a join between two tables.

This snippet works perfectly.
```rust
let mut query: Vec<User> = users::table
   .left_join(sectors::table)
   .group_by(users::uuid)
   .select(users::all_columns)
   .filter((sectors::id.eq("foo")))
   .load(&conn);
```

But when I try to create a function to dynamically construct a filter, I get a lot of errors.
```rust
pub fn user_filter() -> Box<
    dyn BoxableExpression<
        LeftJoinQuerySource<users::table, sectors::table>,
        DB,
        SqlType = Nullable<Bool>,
    >,
> {
    Box::new(sectors::id.eq("foo")) // Errors here
}
```

<details>
<summary>Errors</summary>
<pre><code>
error[E0271]: type mismatch resolving `<diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<db::schema::sectors::columns::id, diesel::expression::bound::Bound<diesel::sql_types::Text, &str>>> as diesel::Expression>::SqlType == diesel::sql_types::Nullable<diesel::sql_types::Bool>`
   --> src/graphql/query/queries/users.rs:158:5
    |
158 |     Box::new(sectors::id.eq("foo"))
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `diesel::sql_types::Nullable`, found struct `diesel::sql_types::Bool`
    |
    = note: expected struct `diesel::sql_types::Nullable<diesel::sql_types::Bool>`
               found struct `diesel::sql_types::Bool`
    = note: required for the cast from `diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<db::schema::sectors::columns::id, diesel::expression::bound::Bound<diesel::sql_types::Text, &str>>>` to the object type `dyn diesel::BoxableExpression<JoinOn<query_source::joins::Join<db::schema::users::table, db::schema::sectors::table, LeftOuter>, diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<NullableExpression<db::schema::users::columns::sector_uuid>, NullableExpression<db::schema::sectors::columns::uuid>>>>, Pg, SqlType = diesel::sql_types::Nullable<diesel::sql_types::Bool>>`

error[E0271]: type mismatch resolving `<db::schema::sectors::table as AppearsInFromClause<db::schema::sectors::table>>::Count == diesel::query_source::Never`
   --> src/graphql/query/queries/users.rs:158:5
    |
158 |     Box::new(sectors::id.eq("foo"))
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `diesel::query_source::Never`, found struct `diesel::query_source::Once`
    |
note: required for `db::schema::sectors::columns::id` to implement `SelectableExpression<query_source::joins::Join<db::schema::users::table, db::schema::sectors::table, LeftOuter>>`
   --> src/db/schema.rs:64:1
    |
64  | / diesel::table! {
65  | |     sectors (uuid) {
66  | |         uuid -> Uuid,
67  | |         id -> Varchar,
...   |
71  | |     }
72  | | }
    | |_^
    = note: 3 redundant requirements hidden
    = note: required for `diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<db::schema::sectors::columns::id, diesel::expression::bound::Bound<diesel::sql_types::Text, &str>>>` to implement `SelectableExpression<JoinOn<query_source::joins::Join<db::schema::users::table, db::schema::sectors::table, LeftOuter>, diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<NullableExpression<db::schema::users::columns::sector_uuid>, NullableExpression<db::schema::sectors::columns::uuid>>>>>`
    = note: required for `diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<db::schema::sectors::columns::id, diesel::expression::bound::Bound<diesel::sql_types::Text, &str>>>` to implement `diesel::BoxableExpression<JoinOn<query_source::joins::Join<db::schema::users::table, db::schema::sectors::table, LeftOuter>, diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<NullableExpression<db::schema::users::columns::sector_uuid>, NullableExpression<db::schema::sectors::columns::uuid>>>>, Pg>`
    = note: required for the cast from `diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<db::schema::sectors::columns::id, diesel::expression::bound::Bound<diesel::sql_types::Text, &str>>>` to the object type `dyn diesel::BoxableExpression<JoinOn<query_source::joins::Join<db::schema::users::table, db::schema::sectors::table, LeftOuter>, diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<NullableExpression<db::schema::users::columns::sector_uuid>, NullableExpression<db::schema::sectors::columns::uuid>>>>, Pg, SqlType = diesel::sql_types::Nullable<diesel::sql_types::Bool>>`
    = note: this error originates in the macro `$crate::__diesel_column` which comes from the expansion of the macro `diesel::table` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `db::schema::sectors::columns::id: SelectableExpression<db::schema::users::table>` is not satisfied
   --> src/graphql/query/queries/users.rs:158:5
    |
158 |     Box::new(sectors::id.eq("foo"))
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `SelectableExpression<db::schema::users::table>` is not implemented for `db::schema::sectors::columns::id`
    |
    = help: the following other types implement trait `SelectableExpression<QS>`:
              <db::schema::sectors::columns::id as SelectableExpression<JoinOn<Join, On>>>
              <db::schema::sectors::columns::id as SelectableExpression<Only<db::schema::sectors::table>>>
              <db::schema::sectors::columns::id as SelectableExpression<SelectStatement<FromClause<From>>>>
              <db::schema::sectors::columns::id as SelectableExpression<db::schema::sectors::table>>
              <db::schema::sectors::columns::id as SelectableExpression<query_source::joins::Join<Left, Right, Inner>>>
              <db::schema::sectors::columns::id as SelectableExpression<query_source::joins::Join<Left, Right, LeftOuter>>>
note: required for `db::schema::sectors::columns::id` to implement `SelectableExpression<query_source::joins::Join<db::schema::users::table, db::schema::sectors::table, LeftOuter>>`
   --> src/db/schema.rs:64:1
    |
64  | / diesel::table! {
65  | |     sectors (uuid) {
66  | |         uuid -> Uuid,
67  | |         id -> Varchar,
...   |
71  | |     }
72  | | }
    | |_^
    = note: 3 redundant requirements hidden
    = note: required for `diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<db::schema::sectors::columns::id, diesel::expression::bound::Bound<diesel::sql_types::Text, &str>>>` to implement `SelectableExpression<JoinOn<query_source::joins::Join<db::schema::users::table, db::schema::sectors::table, LeftOuter>, diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<NullableExpression<db::schema::users::columns::sector_uuid>, NullableExpression<db::schema::sectors::columns::uuid>>>>>`
    = note: required for `diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<db::schema::sectors::columns::id, diesel::expression::bound::Bound<diesel::sql_types::Text, &str>>>` to implement `diesel::BoxableExpression<JoinOn<query_source::joins::Join<db::schema::users::table, db::schema::sectors::table, LeftOuter>, diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<NullableExpression<db::schema::users::columns::sector_uuid>, NullableExpression<db::schema::sectors::columns::uuid>>>>, Pg>`
    = note: required for the cast from `diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<db::schema::sectors::columns::id, diesel::expression::bound::Bound<diesel::sql_types::Text, &str>>>` to the object type `dyn diesel::BoxableExpression<JoinOn<query_source::joins::Join<db::schema::users::table, db::schema::sectors::table, LeftOuter>, diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<NullableExpression<db::schema::users::columns::sector_uuid>, NullableExpression<db::schema::sectors::columns::uuid>>>>, Pg, SqlType = diesel::sql_types::Nullable<diesel::sql_types::Bool>>`
    = note: this error originates in the macro `$crate::__diesel_column` which comes from the expansion of the macro `diesel::table` (in Nightly builds, run with -Z macro-backtrace for more info)
</code></pre>
</details>

Simplified version of `schema.rs` :
```rust
diesel::table! {
    users (uuid) {
        uuid -> Uuid,
        id -> Varchar,
        // ...
        sector_uuid -> Uuid,
    }
}

diesel::table! {
    sectors (uuid) {
        uuid -> Uuid,
        id -> Varchar,
        // ...
    }
}

diesel::joinable!(users -> sectors (sector_uuid));

diesel::allow_tables_to_appear_in_same_query!(
    sectors,
    users,
    // ...
);
```

Thanks !

**Comment by weiznich** [comment, marked as answer]
This is caused by the fact that the expression `sectors::id.eq("foo")` returns a not nullable boolean. Your return type specifies a nullable boolean. You can fix that problem by:
* Using `SqlType = Bool` as return type
* Calling `.nulable()` onto the expression to convert it into a nullable expression, so `Box::new(sectors::id.eq("foo").nullable())`

Depending on your usecase one or the other solution might be preferable. Go with the second one if you add several potential nullable expressions, use the first one if all expressions are guaranteed to be not null.

**Reply by SteelAlloy** [reply]
Thank you very much, I should have seen it earlier. Diesel errors are always quite cryptic.

**Reply by SteelAlloy** [reply]
@weiznich I have another question related to this

The final goal being to filter on several tables, I don't see how to do it with nested joins. The following works fine :

```rust
let mut query = users::table
    .left_join(sectors::table)
    .left_join(club_members::table)
    .group_by(users::uuid)
    .select(selection)
    .limit(limit)
    .into_boxed();

query = query.filter(user_filter()); // OK
```

Following the documentation, I can achieve consecutive joins with the following syntax :

```rust
pub fn user_filter() -> Box<
    dyn BoxableExpression<
        LeftJoinQuerySource<LeftJoinQuerySource<users::table, sectors::table>, club_members::table>, // OK
        DB,
        SqlType = Nullable<Bool>,
    >,
> {
    Box::new(sectors::id.eq("foo").nullable())
}
```

But for nested joints, I can't find anywhere how to do it. By nested joins, I mean the following :
```rust
let mut query = users::table
    .left_join(club_members::table.left_join(clubs::table))
    .group_by(users::uuid)
    .select(selection)
    .limit(limit)
    .into_boxed();
    
    query = query.filter(user_filter()); // Error
```

The type that comes to my mind is `LeftJoinQuerySource<users::table, LeftJoinQuerySource<club_members::table, clubs::table>>`, but that doesn't seem to please the compiler. I think I did all the possible combinations during my tests, I guess I need something else.

<details>
<summary>Error</summary>
<pre><code>
error[E0277]: the trait bound `dyn diesel::BoxableExpression<JoinOn<query_source::joins::Join<db::schema::users::table, JoinOn<query_source::joins::Join<db::schema::club_members::table, db::schema::clubs::table, LeftOuter>, diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<NullableExpression<db::schema::club_members::columns::club_uuid>, NullableExpression<db::schema::clubs::columns::uuid>>>>, LeftOuter>, diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<NullableExpression<db::schema::club_members::columns::user_uuid>, NullableExpression<db::schema::users::columns::uuid>>>>, Pg, SqlType = diesel::sql_types::Nullable<diesel::sql_types::Bool>>: AppearsOnTable<JoinOn<query_source::joins::Join<db::schema::users::table, SelectStatement<FromClause<JoinOn<query_source::joins::Join<db::schema::club_members::table, db::schema::clubs::table, LeftOuter>, diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<NullableExpression<db::schema::club_members::columns::club_uuid>, NullableExpression<db::schema::clubs::columns::uuid>>>>>>, LeftOuter>, diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<NullableExpression<db::schema::club_members::columns::user_uuid>, NullableExpression<db::schema::users::columns::uuid>>>>>` is not satisfied
   --> src/graphql/query/queries/users.rs:214:35
    |
214 |                     query = query.filter(user_filter(filter));
    |                                   ^^^^^^ the trait `AppearsOnTable<JoinOn<query_source::joins::Join<db::schema::users::table, SelectStatement<FromClause<JoinOn<query_source::joins::Join<db::schema::club_members::table, db::schema::clubs::table, LeftOuter>, diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<NullableExpression<db::schema::club_members::columns::club_uuid>, NullableExpression<db::schema::clubs::columns::uuid>>>>>>, LeftOuter>, diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<NullableExpression<db::schema::club_members::columns::user_uuid>, NullableExpression<db::schema::users::columns::uuid>>>>>` is not implemented for `dyn diesel::BoxableExpression<JoinOn<query_source::joins::Join<db::schema::users::table, JoinOn<query_source::joins::Join<db::schema::club_members::table, db::schema::clubs::table, LeftOuter>, diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<NullableExpression<db::schema::club_members::columns::club_uuid>, NullableExpression<db::schema::clubs::columns::uuid>>>>, LeftOuter>, diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<NullableExpression<db::schema::club_members::columns::user_uuid>, NullableExpression<db::schema::users::columns::uuid>>>>, Pg, SqlType = diesel::sql_types::Nullable<diesel::sql_types::Bool>>`
    |
    = help: the following other types implement trait `AppearsOnTable<QS>`:
              <&'a T as AppearsOnTable<QS>>
              <(T0, T1) as AppearsOnTable<QS>>
              <(T0, T1, T2) as AppearsOnTable<QS>>
              <(T0, T1, T2, T3) as AppearsOnTable<QS>>
              <(T0, T1, T2, T3, T4) as AppearsOnTable<QS>>
              <(T0, T1, T2, T3, T4, T5) as AppearsOnTable<QS>>
              <(T0, T1, T2, T3, T4, T5, T6) as AppearsOnTable<QS>>
              <(T0, T1, T2, T3, T4, T5, T6, T7) as AppearsOnTable<QS>>
            and 202 others
    = note: required for `Box<dyn diesel::BoxableExpression<JoinOn<query_source::joins::Join<db::schema::users::table, JoinOn<query_source::joins::Join<db::schema::club_members::table, db::schema::clubs::table, LeftOuter>, diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<NullableExpression<db::schema::club_members::columns::club_uuid>, NullableExpression<db::schema::clubs::columns::uuid>>>>, LeftOuter>, diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<NullableExpression<db::schema::club_members::columns::user_uuid>, NullableExpression<db::schema::users::columns::uuid>>>>, Pg, SqlType = diesel::sql_types::Nullable<diesel::sql_types::Bool>>>` to implement `AppearsOnTable<JoinOn<query_source::joins::Join<db::schema::users::table, SelectStatement<FromClause<JoinOn<query_source::joins::Join<db::schema::club_members::table, db::schema::clubs::table, LeftOuter>, diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<NullableExpression<db::schema::club_members::columns::club_uuid>, NullableExpression<db::schema::clubs::columns::uuid>>>>>>, LeftOuter>, diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<NullableExpression<db::schema::club_members::columns::user_uuid>, NullableExpression<db::schema::users::columns::uuid>>>>>`
    = note: required for `BoxedSelectStatement<'_, ((diesel::sql_types::Uuid, diesel::sql_types::Text, diesel::sql_types::Text, diesel::sql_types::Text, diesel::sql_types::Text, diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::Nullable<diesel::sql_types::Text>, db::schema::sql_types::Gender, diesel::sql_types::Nullable<diesel::sql_types::Date>, SmallInt, SmallInt, diesel::sql_types::Uuid, diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::Text, diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::Bool, diesel::sql_types::Bool, diesel::sql_types::Bool, diesel::sql_types::Bool, diesel::sql_types::Bool), diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::BigInt), FromClause<JoinOn<query_source::joins::Join<db::schema::users::table, SelectStatement<FromClause<JoinOn<query_source::joins::Join<db::schema::club_members::table, db::schema::clubs::table, LeftOuter>, diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<NullableExpression<db::schema::club_members::columns::club_uuid>, NullableExpression<db::schema::clubs::columns::uuid>>>>>>, LeftOuter>, diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<NullableExpression<db::schema::club_members::columns::user_uuid>, NullableExpression<db::schema::users::columns::uuid>>>>>, Pg, db::schema::users::columns::uuid>` to implement `FilterDsl<Box<dyn diesel::BoxableExpression<JoinOn<query_source::joins::Join<db::schema::users::table, JoinOn<query_source::joins::Join<db::schema::club_members::table, db::schema::clubs::table, LeftOuter>, diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<NullableExpression<db::schema::club_members::columns::club_uuid>, NullableExpression<db::schema::clubs::columns::uuid>>>>, LeftOuter>, diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<NullableExpression<db::schema::club_members::columns::user_uuid>, NullableExpression<db::schema::users::columns::uuid>>>>, Pg, SqlType = diesel::sql_types::Nullable<diesel::sql_types::Bool>>>>`

</code></pre>
</details>

Missing parts of `schema.rs`

```rust
diesel::table! {
    club_members (user_uuid, club_uuid) {
        user_uuid -> Uuid,
        club_uuid -> Uuid,
        // ...
    }
}

diesel::table! {
    clubs (uuid) {
        uuid -> Uuid,
        id -> Varchar,
        // ...
    }
}

diesel::joinable!(club_members -> clubs (club_uuid));
diesel::joinable!(club_members -> users (user_uuid));

diesel::allow_tables_to_appear_in_same_query!(
    club_members,
    clubs,
    sectors,
    users,
);
```

Thank you for your time ❤️

**Reply by weiznich** [reply]
It's a bit surprising but if you pass an join to a join as type argument you need to use `dsl::LeftJoin` and not `dsl::LeftJoinQuerySource` for the inner join type.

So for reference the following filter compiles:

```rust

pub fn user_filter() -> Box<
    dyn BoxableExpression<
        diesel::dsl::LeftJoinQuerySource<
            users::table,
            diesel::dsl::LeftJoin<club_members::table, clubs::table>,
        >,
        diesel::pg::Pg,
        SqlType = diesel::sql_types::Nullable<diesel::sql_types::Bool>,
    >,
> {
    // obviously needs to use something that's valid for the set of joined tables.
    Box::new(users::id.eq("whatever").nullable()) 
}

fn main() {
    // using group by here would complicate things even more
    // the boxableexpressions docs have some details about that.
    let mut query = users::table
        .left_join(club_members::table.left_join(clubs::table))
        .into_boxed();

    query = query.filter(user_filter()); 
}

```
